File: src/Dice.java

/**
 * Created by arianfarahani on 4/12/17.
 */
import java.util.concurrent.ThreadLocalRandom;

public class Dice
{
      private Dice()
      {
      }

      private static int rollDice(int count, int sides) {
            int total = 0;
            for(int i = 0; i < count; i++) {
                  total += ThreadLocalRandom.current().nextInt(1, sides + 1);
            }
            return total;
      }

      private static int rollRange(int start, int end) {
            return ThreadLocalRandom.current().nextInt(start, end + 1);
      }

      public static int roll(String input) {
            int bonus = 0;
            String base;
            if(input.contains("+")) {
                  String[] mod = input.split("\\+");
                  bonus += Integer.parseInt(mod[1]);
                  base = mod[0];
            } else if(input.contains("-")) {
                  String[] mod = input.split("-");
                  bonus -= Integer.parseInt(mod[1]);
                  base = mod[0];
            } else {
                  base = input;
            }

            String[] dice = base.split("d");

            int count = Integer.parseInt(dice[0]);
            int sides = Integer.parseInt(dice[1]);
            int result = rollDice(count, sides);
            result += bonus;

            //always min 1
            if(result < 1)
                  result = 1;

            return result;
      }

      public static int roll(int input) {
            return rollDice(1, input);
      }

      public static int roll(int start, int end) {
            return rollRange(start, end);
      }

}

File: src/com/leftyist/kingdomsim/utils/Dice.kt
package com.leftyist.kingdomsim.utils;

import java.util.concurrent.ThreadLocalRandom

private fun recursiveParse(input: String): Int {

      val plus = input.indexOf("+")
      val minus = input.indexOf("-")

      var splitPoint = 0

      if (minus == -1 && plus == -1)
            return parsePiece(input)
      else if (minus != -1 && plus != -1)
            splitPoint = minOf(minus, plus)
      else
            splitPoint = maxOf(minus, plus)

      val left = input.substring(0, splitPoint)
      val right = input.substring(splitPoint + 1, input.length)

      if (input[splitPoint] == '+')
            return parsePiece(left) + recursiveParse(right)
      else
            return parsePiece(left) - recursiveParse(right)
}

private fun parsePiece(input: String): Int {
      if (input.contains("d")) {
            return rollDice(input);
      } else {
            return input.toInt();
      }
}

private fun rollDice(dice: String): Int {
      val halves = dice.split("d")

      var sum = 0;
      repeat(halves[0].toInt()) {
            sum += rollRange(max = halves[1].toInt())
      }
      return sum;
}

fun rollRange(min: Int = 1, max: Int): Int {
      return ThreadLocalRandom.current().nextInt(min, max + 1)
}

fun roll(roll: String): Int {
      val result = recursiveParse(roll.replace(" ", ""))
      if (result < 1)
            return 1
      else
            return result
}